File: src/main/java/com/ultreon/ultranlang/token/TokenType.java
package com.ultreon.ultranlang.token;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.*;

@SuppressWarnings("SpellCheckingInspection")
public record TokenType(@NotNull String value) {
    private static final Map<String, TokenType> VALUE_TO_INSTANCE = new HashMap<>();
    private static final List<TokenType> INSTANCES = new ArrayList<>();

    // single-character token types
    public static final TokenType PLUS = new TokenType("+");
    public static final TokenType MINUS = new TokenType("-");
    public static final TokenType MUL = new TokenType("*");
    public static final TokenType FLOAT_DIV = new TokenType("/");
    public static final TokenType LPAREN = new TokenType("(");
    public static final TokenType RPAREN = new TokenType(")");
    public static final TokenType LCURL = new TokenType("{");
    public static final TokenType RCURL = new TokenType("}");
    public static final TokenType SEMI = new TokenType(";");
    public static final TokenType DOT = new TokenType(".");
    public static final TokenType COLON = new TokenType(":");
    public static final TokenType COMMA = new TokenType(",");

    // block of reserved words
    public static final TokenType PROGRAM = new TokenType("PROGRAM");
    public static final TokenType INTEGER = new TokenType("INTEGER");
    public static final TokenType REAL = new TokenType("REAL");
    public static final TokenType STRING = new TokenType("STRING");
    public static final TokenType BOOLEAN = new TokenType("BOOLEAN");
    public static final TokenType TRUE = new TokenType("TRUE");
    public static final TokenType FALSE = new TokenType("FALSE");
    public static final TokenType INTEGER_DIV = new TokenType("DIV");
    public static final TokenType VAR = new TokenType("VAR");
    public static final TokenType FUNCTION = new TokenType("FUNCTION");
    public static final TokenType BEGIN = new TokenType("BEGIN");
    public static final TokenType END = new TokenType("END");
    public static final TokenType RETURN = new TokenType("RETURN");

    // misc
    public static final TokenType ID = new TokenType("ID");
    public static final TokenType INTEGER_CONST = new TokenType("INTEGER_CONST");
    public static final TokenType REAL_CONST = new TokenType("REAL_CONST");
    public static final TokenType STRING_CONST = new TokenType("INTEGER_CONST");
    public static final TokenType ASSIGN = new TokenType("ASSIGN");
    public static final TokenType EOF = new TokenType("EOF");

    public TokenType(@NotNull String value) {
        this.value = value;
        VALUE_TO_INSTANCE.put(value, this);
        INSTANCES.add(this);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        TokenType tokenType = (TokenType) o;
        return value.equals(tokenType.value);
    }

    @Override
    public int hashCode() {
        return value.hashCode();
    }

    @Override
    public String toString() {
        return value;
    }

    @Nullable
    public static TokenType fromString(String value) {
        return VALUE_TO_INSTANCE.get(value);
    }

    public static List<TokenType> getTypes() {
        return Collections.unmodifiableList(INSTANCES);
    }
}

File: src/main/kotlin/com/ultreon/ultranlang/token/TokenType.kt
package com.ultreon.ultranlang.token

@Suppress("SpellCheckingInspection")
class TokenType(val value: String) {
    override fun toString(): String {
        return value
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true

        other as TokenType

        if (value != other.value) return false

        return true
    }

    val isNewline: Boolean get() = this == SEMI || this == NEW_LINE

    override fun hashCode(): Int {
        return value.hashCode()
    }

    companion object {
        // single-character token types
        val PLUS = TokenType("+")
        val MINUS = TokenType("-")
        val MUL = TokenType("*")
        val FLOAT_DIV = TokenType("/")
        val LPAREN = TokenType("(")
        val RPAREN = TokenType(")")
        val LCURL = TokenType("{")
        val RCURL = TokenType("}")
        val SEMI = TokenType(";")
        val DOT = TokenType(".")
        val COLON = TokenType(":")
        val COMMA = TokenType(",")
        val NEW_LINE = TokenType("\n")

        // block of reserved words
        val PROGRAM = TokenType("PROGRAM")
        val INTEGER = TokenType("INTEGER")
        val REAL = TokenType("REAL")
        val STRING = TokenType("STRING")
        val BOOLEAN = TokenType("BOOLEAN")
        val TRUE = TokenType("TRUE")
        val FALSE = TokenType("FALSE")
        val INTEGER_DIV = TokenType("DIV")
        val VAR = TokenType("VAR")
        val VAL = TokenType("VAL")
        val FUNCTION = TokenType("FUNCTION")
        val CLASS = TokenType("CLASS")
        val CONSTRUCTOR = TokenType("CONSTRUCTOR")
        val STATIC = TokenType("STATIC")
        val BEGIN = TokenType("BEGIN")
        val END = TokenType("END")
        val RETURN = TokenType("RETURN")

        // misc
        val ID = TokenType("ID")
        val INTEGER_CONST = TokenType("INTEGER_CONST")
        val REAL_CONST = TokenType("REAL_CONST")
        val STRING_CONST = TokenType("INTEGER_CONST")
        val ASSIGN = TokenType("ASSIGN")
        val EOF = TokenType("EOF")

        fun values(): Array<TokenType> {
            return arrayOf(PLUS, MINUS, MUL, FLOAT_DIV, LPAREN, RPAREN, LCURL, RCURL, SEMI, NEW_LINE, DOT, COLON, COMMA, PROGRAM,
                INTEGER, REAL,
                INTEGER_DIV, VAR, FUNCTION, BEGIN, END, ID, INTEGER_CONST, STRING_CONST, REAL_CONST, ASSIGN, EOF)
        }

        fun valueOf(value: String): TokenType {
            return when (value) {
                "PLUS" -> PLUS
                "MINUS" -> MINUS
                "MUL" -> MUL
                "FLOAT_DIV" -> FLOAT_DIV
                "LPAREN" -> LPAREN
                "RPAREN" -> RPAREN
                "SEMI" -> SEMI
                "NEW_LINE" -> NEW_LINE
                "DOT" -> DOT
                "COLON" -> COLON
                "COMMA" -> COMMA
                "PROGRAM" -> PROGRAM
                "INTEGER" -> INTEGER
                "REAL" -> REAL
                "INTEGER_DIV" -> INTEGER_DIV
                "VAR" -> VAR
                "FUNCTION" -> FUNCTION
                "CLASS" -> CLASS
                "CONSTRUCTOR" -> CONSTRUCTOR
                "STATIC" -> STATIC
                "BEGIN" -> BEGIN
                "END" -> END
                "ID" -> ID
                "INTEGER_CONST" -> INTEGER_CONST
                "REAL_CONST" -> REAL_CONST
                "ASSIGN" -> ASSIGN
                "EOF" -> EOF
                else -> throw IllegalArgumentException("No object com.ultreon.ultranlang.TokenType.$value")
            }
        }
    }
}